crates/sui-node/src/lib.rs

```diff
@@ -1590,41 +1594,101 @@ impl SuiNode {
-    // TODO: Add comments explain why we need to do this at start up.
-    async fn reenqueue_pending_consensus_certs(
+    /// Re-executes pending consensus certificates, which may not have been committed to disk
+    /// before the node restarted. This is necessary for the following reasons:
+    ///
+    /// 1. For any transaction for which we returned signed effects to a client, we must ensure
+    ///    that we have re-executed the transaction before we begin accepting grpc requests.
+    ///    Otherwise we would appear to have forgotten about the transaction.
+    /// 2. While this is running, we are concurrently waiting for all previously built checkpoints
+    ///    to be rebuilt. Since there may be dependencies in either direction (from checkpointed
+    ///    consensus transactions to pending consensus transactions, or vice versa), we must
+    ///    re-execute pending consensus transactions to ensure that both processes can complete.
+    /// 3. Also note that for any pending consensus transactions for which we wrote a signed effects
+    ///    digest to disk, we must re-execute using that digest as the expected effects digest,
+    ///    to ensure that we cannot arrive at different effects than what we previously signed.
+    async fn reexecute_pending_consensus_certs(
         epoch_store: &Arc<AuthorityPerEpochStore>,
         state: &Arc<AuthorityState>,
     ) {
-        let pending_consensus_certificates = epoch_store
-            .get_all_pending_consensus_transactions()
-            .into_iter()
-            .filter_map(|tx| match tx.kind {
-                ConsensusTransactionKind::CertifiedTransaction(tx) => {
-                    if tx.contains_shared_object() {
-                        // We cannot schedule shared object transactions for execution here
-                        // because they must come out of consensus to get prover version assignment.
-                        None
+        let mut pending_consensus_certificates = Vec::new();
+        let mut additional_certs = Vec::new();
+
+        for tx in epoch_store.get_all_pending_consensus_transactions() {
+            match tx.kind {
+                // shared object txns will be re-executed by consensus replay
+                ConsensusTransactionKind::CertifiedTransaction(tx)
+                    if !tx.contains_shared_object() =>
+                {
+                    let tx = *tx;
+                    // new_unchecked is safe because we never submit a transaction to consensus
+                    // without verifying it
+                    let tx = VerifiedExecutableTransaction::new_from_certificate(
+                        VerifiedCertificate::new_unchecked(tx),
+                    );
+                    // we only need to re-execute if we previously signed the effects (which indicates we
+                    // returned the effects to a client).
+                    if let Some(fx_digest) = epoch_store
```

Comment:
I don't think this is necessary, because of this line: https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/authority.rs#L1281
So I think you could just keep this part unchanged.

Reply:
Oh, I forgot about that piece. I think it would actually be better to remove the other code and keep this, since it removes a db read from the hot path

Reply:
I don't remember but I wonder are there other reasons we had that code?

Reply:
Based on reviewing #18426 I don't think there was any larger reason to do it that way.

Unfortunately it turns out to be quite tricky to remove it while still covering all the cases. For instance, detecting equivocation for replayed shared object transactions that have signed effects, but are not yet checkpointed is trivial with the check in authority.rs but much harder to do otherwise.

I may return to this if this read appears to be a performance problem at some point.